### ISSUE TITLE
fix(model): instrument the V4.1 layer-36 q/kv producer chain so the ~1e18 upstream magnitude source is named, not silently propagated

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13291]
﻿

# fix(model): instrument the V4.1 layer-36 q/kv producer chain so the ~1e18 upstream magnitude source is named, not silently propagated

```routing
lane: model
paths: ["internal/model/v41_attention.go", "internal/model/v41/v41_sparse_attention.go", "internal/model/v41_attention_test.go"]
expected_steps: 3
priority: P1
domain: model
scope_class: leaf
repo: fak
```

## Problem frame

- Centrality: Core - the first (and so far only) wall between the native V4.1 forward and its first token.
- P1 Context: advanced - #13290 landed a fail-closed output guard naming the TRUE producing stage, but explicitly left the upstream magnitude source unexplained.
- P2 Net value: preserved - instrumentation is additive error-path only; finite-input arithmetic stays byte-identical.
- P3 Adaptation: advanced - makes a silent blow-up self-attributing at its origin instead of surfacing two stages later.
- P4 Operations: preserved - no roofline gate touched; pinned vcruz Q2_K `58d8ac86` artifact unchanged; no external-runtime fallback.

## Working spine

The #13290 attribution (`c245be2fa`) converted a silent NaN into a named refusal at the 512-term attention contraction, but its own closing message records the open gap:

> "The remaining frontier is the upstream magnitude source: the guard converts a silent NaN propagation into a named refusal but does not yet explain why O(1)-scale activations reach ~1e18 entering the layer-36 sink."

Name that origin so the next physical run attributes the blow-up instead of re-deriving it.

## Current state

Layer 36 on the published schedule has `Ratio=1`, resolves `KVSourceLayer=20` (also ratio 1), so `sharedKV` stays nil and layer 36 executes the per-position `V41SparseAttentionSink` path. Both V4.1 contractions now validate their OUTPUT (typed `ErrV41SparseSinkNonFinite` / `v41CompressedNonFinite`), but the INPUTS `q`/`kv` are only checked for finiteness - not for an out-of-envelope magnitude. A finite per-element magnitude above `sqrt(MaxFloat32/512) ~= 8.152e17` still overflows the 512-term `dot += q*kv` accumulation. The layer-36 q/kv producer chain is:

- `q`: `attn.wq_a.weight` (`QLoraRank`) -> `attn.wq_b.weight` (`nH*hd`), then `applyRopeTailInterleaved(q[h*hd:], cos, sin, qk_rope_head_dim)`.
- `kv`: `attn.wkv.weight` (`v41KVLoraRank`, sliced to `hd`), then `applyRopeTailInterleaved(kv, cos, sin, qk_rope_head_dim)` (`internal/model/v41_forward.go` ~1702-1740).

`applyRopeTailInterleaved` (`internal/model/rope.go:234`) is an orthonormal adjacent-pair rotation scaled by `cfg.ropeAttentionFactor()`; it cannot manufacture ~1e18 from O(1) inputs, so the blow-up is upstream in the projection output or the table scale. That seam is currently unnamed.

## Core through-line

Instrument the layer-36 q/kv producer chain (projection output magnitude, then the table scale) -> a finite-but-out-of-envelope magnitude at the true origin is refused with a typed token naming layer/head/position/element -> a RED->GREEN regression pins it -> the next physical strix3 run attributes the blow-up directly.

## Gold-plating boundary

No change to attention geometry, RoPE convention, softmax scaling, or any finite-input arithmetic. No new tolerance is invented; the guard names a producer, it does not clamp, rescale, or "fix" a value. No roofline gate lowered; no change to the pinned vcruz Q2_K `58d8ac86` artifact; no external-runtime fallback.

## Done condition / witness

The layer-36 q/kv producer chain refuses (or names) a finite-but-out-of-envelope magnitude at its true origin with a typed error, and the regression passes.
Witness: `go test ./internal/model/... -run TestV41 -count=1`

## Definition of done

- [ ] RED-on-parent regression reproduces an out-of-envelope q/kv magnitude entering the layer-36 sink without naming its origin
- [ ] GREEN: the producing stage is named with a typed token (layer/head/position/element)
- [ ] `go test ./internal/model/...` PASS; `go vet ./internal/model ./internal/model/v41` clean; 5-gate boundary PASS
- [ ] Finite-input arithmetic byte-identical; existing `TestV41*` suites stay green

## Acceptance gate

`go test ./internal/model/... -count=1` PASS plus `go vet` clean plus 5-gate boundary PASS.

## Closure binding

Resolving commit cites this issue and carries the `(fak model)` trailer.

## Lane

model

## Parent context

anthony-chaudhary/fak#12640 (native V4.1 Flash execution and physical qualification); predecessor anthony-chaudhary/fak#13290 (landed `c245be2fa`).

## Likely files

- internal/model/v41_forward.go (producer chain, ~1702-1740)
- internal/model/v41/v41_sparse_attention.go
- internal/model/v41_attention.go
- internal/model/rope.go (`applyRopeTailInterleaved` / `v41RopeTableForLayer`)

## Expected steps

3

- Centrality: Core
- P1 Context: advanced
- P2 Net value: preserved
- P3 Adaptation: advanced
- P4 Operations: preserved

## Overall completion contribution

Contribution: 2/40 points on the native V4.1 forward spine (fak#12640).

## Provenance

- Public trunk at filing: `72172454f`.
- Predecessor receipt: `.fak/ops-receipts/mission-native/2026-09-19T0302Z-c245be2fa-layer36-nonfinite-names-true-producer.md`.
- No native V4.1 generation or physical parity claimed.

**Comment by anthony-chaudhary** [comment]
## [SW-VERIFIED] The layer-36 magnitude source is NAMED and FIXED — missing V4.1 Q/KV latent norms; landed `a03df5a8e` on public `origin/main`

**Frontier step (arm A, code leaf).** This issue was filed to *instrument* the ~1e18 upstream magnitude source feeding the layer-36 attention contraction. The root cause was found directly and fixed at the source instead: the native full-path V4.1 forward never applied the reference Attention's two latent RMSNorms.

### Root cause (the magnitude source, named)

The reference `inference/model.py` `Attention.forward` is explicit:

```python
qr = self.q_norm(self.wq_a(x))      # self.q_norm = RMSNorm(q_lora_rank)
q  = self.wq_b(qr)
kv = self.kv_norm(self.wkv(x))      # self.kv_norm = RMSNorm(head_dim), in _window_kv
apply_rotary_emb(kv[..., -rd:], freqs_cis)
```

The native forward projected `attn.wq_a` and `attn.wkv` and went straight to RoPE with **no RMSNorm on either latent**. The loader had already mapped the two tensors to dedicated leaves (`attn.wq_a_norm.weight`, `attn.kv_norm.weight`) and `internal/ggufload/deepseek41.go` documented them as "deliberately NOT wired … wiring them into the forward is a separate leaf". That separate leaf is this one.

Without the norms, the Q and KV vectors reaching the 512-term attention dot product are raw unbounded projections: a finite per-element magnitude above `sqrt(MaxFloat32/512) ~= 8.152e17` overflows `dot += q*kv` to `+Inf`; `maxScore` becomes `+Inf`; `exp32(Inf-Inf)=NaN` poisons the softmax denominator (the `sum == 0` guard does not fire on NaN); and the NaN row surfaces two stages later at `V41GroupedOutputProjection` as the layer-36 refusal `model: V4.1 grouped output non-finite value element 0` (fak#13290). This is exactly the "O(1)-scale activations reach ~1e18 entering the layer-36 sink" gap.

### Fix (`a03df5a8e`)

Wire both norms on the **full path only** (the reduced fixture's pre-#13009 arithmetic stays byte-identical), matching the reference order:

- `q_norm` over the q-lora latent before `attn.wq_b`;
- `kv_norm` over the KV vector (full `head_dim`) before the rope tail.

Admission of `attn.wq_a_norm.weight` (`QLoraRank`) and `attn.kv_norm.weight` (`v41KVLoraRank`) is gated on the full geometry and fails closed by name when absent. No attention geometry, RoPE convention, softmax scaling, or finite-input arithmetic changed; no roofline gate touched; the pinned vcruz Q2_K `58d8ac86` artifact is unchanged.

### Witness

- **RED-on-parent reproduced the identical non-finite class.** With `attn.wq_a`/`wq_b`/`wkv` scaled to 1e18, the parent forward fails: `model: V4.1 forward stage attention layer=0: model: V4.1 grouped output non-finite value element 0`. GREEN after.
- New `TestV41AttnLatentNormAppliedOnFullPath` pins both stages against an independent scalar oracle (RMSNorm-then-project order), proves the construction is discriminating (un-normed ≠ normed), reproduces the 512-term overflow directly, and drives the real `forwardV41` on the full-geometry fixture to finite logits.
- `go test ./internal/model/ -count=1` → **ok** (180.8s); `go test ./internal/ggufload/ -count=1` → **ok** (15.5s); `go vet ./internal/model/ ./internal/ggufload/` clean; `gofmt -l` empty.
- 5-Gate boundary **PASS** (5 Go files, 0 violations) via the real pre-commit hook; secret-leak audit clean; DOS `commit-audit a03df5a8e` → **OK / diff-witnessed**.

**Not claimed.** No token generated; no prefill/decode tok/s; the two-prompt box and the 30/5 `[HW-WITNESSED]` first-witness bar remain `[ ]`. This is `[SW-VERIFIED]` only — a software-only code leaf, NOT physical evidence. Not a qualification, not a speedup; no roofline gate lowered.

**Next frontier (one step).** The **physical strix3 re-run at `a03df5a8e`** under a fresh exclusive lease: build the `-tags vulkan` binary from the new trunk and run the same recipe (`FAK_STREAM_Q4K=1 --engine inkernel --backend vulkan --cpu-offload-experts --addr 127.0.0.1:8084`) toward the two-prompt / native-identity / 30/5 first-token boxes. The layer-36 attention input is now bounded; a physical re-run is the only way to tell whether the forward advances past layer 36.

_Landed + pushed `a03df5a8e` (`anthony-chaudhary/fak`). `Closes #13291`; `Ref #13290, #12640`._

**Comment by anthony-chaudhary** [comment]
﻿## [SW-VERIFIED] Run receipt — duplicate leaf WITHHELD; resolved by `a03df5a8e`

This lane independently selected and implemented a **producer-side envelope guard** for this exact frontier (`v41AttentionProducerEnvelopeCheck`: refuse a finite attention operand whose magnitude exceeds `sqrt(MaxFloat32/HeadDim)` at HeadDim=512, at the true producing stage), then discovered a peer had already resolved this issue at the ROOT:

- **`a03df5a8e`** `fix(model): apply the V4.1 Q/KV latent norms, naming the layer-36 magnitude source (fak model)` — **Closes #13291**.

The peer's fix is strictly better than a detector: the reference `Attention.forward` applies `q_norm = RMSNorm(q_lora_rank)` over the q-lora latent before `wq_b`, and `kv_norm = RMSNorm(head_dim)` over the KV vector before RoPE; omitting them let raw unbounded projections reach the 512-term dot. That is *why* O(1)-scale activations became ~1e18 — the question this issue was filed to answer.

**Action taken: duplicate withheld, not landed.** This lane's commit stays on a throwaway branch (never on `main`/`origin`); the worktree and branch were removed.

Independent witness on trunk (`fed6a6a37`):
```
go test ./internal/model/ -run TestV41AttnLatentNormAppliedOnFullPath -count=1  => PASS
gh issue view 13291 --json state => CLOSED (closedAt 2026-09-19T10:40:01Z)
git -C fak branch --contains <this-lane-commit> => only fak/ticket-13291 (withheld)
```

No public `main` commit landed by this run; no `[HW-WITNESSED]` criterion touched; #12640 stays OPEN; V4 Flash 0731 control untouched.
